### ISSUE TITLE
Incorrect output when dealing with duplicate property declarations in a selector

### ISSUE THREAD

**Issue by jimmy-guo** [elchininet/postcss-rtlcss#181]
Playground reproduction of the issue: https://elchininet.github.io/postcss-rtlcss/#63dd66e7f0726

Input

```
.test1, .test2 {
    margin-right: 4px;
    margin-right: var(--hello);
}
```

Output

```
.test1, .test2 {
    margin-right: 4px;
}

[dir="ltr"] .test1, [dir="ltr"] .test2 {
    margin-right: var(--hello);
}

[dir="rtl"] .test1, [dir="rtl"] .test2 {
    margin-left: var(--hello);
}
```

Expected output
```
[dir="ltr"] .test1, [dir="ltr"] .test2 {
    margin-right: var(--hello);
}

[dir="rtl"] .test1, [dir="rtl"] .test2 {
    margin-left: var(--hello);
}
```

When I try to RTL a selector with duplicate property declarations, but different values, the library seems smart enough to know that the last declaration applies, but not all instances of the property are removed from the non-prefixed LTR / RTL selectors.

(In our particular use case, we are extending styles of components by merging styles together, resulting in duplicate property declarations with different values, and noticed that the RTL versions still had some LTR styling applied.)

**Comment by elchininet** [comment]
Hi @jimmy-guo,

`postcss-rtlcss` plugin will not remove any declaration or rule from CSS files, its purpose is just to create prefixed rules containing flipped versions. To remove overridden rules or declarations from your CSS you would need to use other plugins that are intended for this, [but even plugins that are intended to do this might not do it in that specific case](https://github.com/cssnano/cssnano/issues/846).

That‘s being said, checking your specific example, the only way that `margin-right: 4px` is applied to your elements, is that there is no `dir` property in your elements‘ parents or because you are testing in a browser that doesn‘t support environment variables so the fallback value is applied.

Could you explain a bit how your prefixed rules are not overriding your non-prefixed rule? Maybe I could come with an idea to overcome this or bring a new option that could cover your use case.

Regards.

**Comment by jimmy-guo** [comment]
Hi @elchininet, thanks for the response. Not sure if "remove" is the best word, but to illustrate what I was imagining, [here](https://elchininet.github.io/postcss-rtlcss/#63e43f0057fa1) is an example input and output with just one `margin-right` declaration: 

Input

```
.test1, .test2 {
    margin-right: var(--hello);
}
```

Output

```
[dir="ltr"] .test1, [dir="ltr"] .test2 {
    margin-right: var(--hello);
}

[dir="rtl"] .test1, [dir="rtl"] .test2 {
    margin-left: var(--hello);
}
```

It seems to "remove" the original  

```
.test1, .test2 {
    margin-right: var(--hello);
}
```
in favor of the RTL and LTR selectors for this style, and I was expecting the same to happen even when there are multiple declarations for the same property.

The output in my original description is problematic because in RTL, both `margin-right: 4px` and `margin-left: var(--hello)` apply. However, `margin-right: 4px` should only apply in LTR languages (and be correctly overriden by `margin-right: var(--hello)`).

```
.test1, .test2 {
    margin-right: 4px;
}

[dir="ltr"] .test1, [dir="ltr"] .test2 {
    margin-right: var(--hello);
}

[dir="rtl"] .test1, [dir="rtl"] .test2 {
    margin-left: var(--hello);
}
```

Hope that helps clarify my thoughts!

**Comment by elchininet** [comment]
Hi @jimmy-guo,

That‘s right, what is removed in your first example is an empty rule (the declaration has been moved to the prefixed RTL rule and the original rule ended up empty). This is something that was introduced in early versions of the plugin to avoid ending up with empty rules. But if the rule contains any declaration, like in your second case, it is not removed.

I understand that when you say _This output is problematic_ is because you want your CSS to be used in browsers that don‘t support CSS variables, otherwise it is not problematic in any way. In a browser that supports CSS variables, it doesn‘t matter if the un-prefixed CSS rule is there, it will not be applied even if the variable is not defined, it will be always overridden by the prefixed one.

On the other hand, it is true that for backward compatibility, specifically to support Internet Explorer, all the declarations that contains `var`, `env`, `calc`, etc should be moved to the prefixed version, and the first overridden declaration that doesn‘t contain any function should be moved too, but for the rest it should remain as it is to avoid increasing unnecessarily the size of the final bundle.

So, this is my proposal that could be introduced with a new option:

### 1) If no CSS functions are used it should remain as it is right now

##### Input

```

.test {
    margin-right: 1px;
    margin-right: 2px;
    margin-right: 3px;
    margin-right: 4px;
    margin-right: 5px;
}
```

##### Output

```
.test {
    margin-right: 1px;
    margin-right: 2px;
    margin-right: 3px;
    margin-right: 4px;
}

[dir="ltr"] .test {
    margin-right: 5px;
}

[dir="rtl"] .test {
    margin-left: 5px;
}
```

The right margins `1px`, `2px`, `3px` and `4px` will be present in your RTL elements but they will be always overridden by the `5px` one.

### 2) If only one CSS function is used:

##### Input

```

.test {
    margin-right: 1px;
    margin-right: 2px;
    margin-right: 3px;
    margin-right: 4px;
    margin-right: var(--hello);
}
```

##### Output

```
.test {
    margin-right: 1px;
    margin-right: 2px;
    margin-right: 3px;
}

[dir="ltr"] .test {
    margin-right: 4px;
    margin-right: var(--hello);
}

[dir="rtl"] .test {
    margin-left: 4px;
    margin-left: var(--hello);
}
```

Only the first overridden declaration that doesn‘t use any CSS function should be moved, so if the browser doesn‘t support the CSS function the fallback value should be used. In this case occurs the same, right margins `1px`, `2px`, and `3px` will be present in your RTL rules, but they will be always overridden.

### 2) Multiple CSS functions are used:

```

.test {
    margin-right: 1px;
    margin-right: 2px;
    margin-right: constant(--hello);
    margin-right: env(--hello);
    margin-right: var(--hello);
}
```

##### Output

```
.test {
    margin-right: 1px;   
}

[dir="ltr"] .test {
    margin-right: 2px;
    margin-right: constant(--hello);
    margin-right: env(--hello);
    margin-right: var(--hello);
}

[dir="rtl"] .test {
    margin-left: 2px;
    margin-left: constant(--hello);
    margin-left: env(--hello);
    margin-left: var(--hello);
}
```

All the CSS functions should be moved and only the first overridden rule that doesn‘t contain any CSS function should be moved. Again, right margin `1px` will be present in your RTL elements but always overridden.

Let me know if this solves your use case and I‘ll implement the new option when I got some time.

Regards.

**Comment by elchininet** [comment]
Hi @jimmy-guo,

Now I think that I understood your issue, `margin-right` will not be overridden because in RTL it is a `margin-left`. I think that in these kinds of override all the rules should be moved. Let me come up with a solution. I'll let you now.

Regards.

**Comment by elchininet** [comment]
@jimmy-guo,
I deleted my previous proposal because it doesn‘t make sense to implement it. I‘ll let you know when I come up with a solution for your use case.
Regards.

**Comment by jimmy-guo** [comment]
> Hi @jimmy-guo,
> 
> Now I think that I understood your issue, `margin-right` will not be overridden because in RTL it is a `margin-left`. I think that in these kinds of override all the rules should be moved. Let me come up with a solution. I'll let you now.
> 
> Regards.

Yes, exactly! My inclusion of a CSS variable overcomplicated what I was trying to illustrate. Just to clarify further, here is a simpler example of the issue:

Input
```
.test1, .test2 {
    margin-right: 4px;
    margin-right: 5px;
}
```

Output

```
.test1, .test2 {
    margin-right: 4px; // will not be overridden and still apply in RTL
}

[dir="ltr"] .test1, [dir="ltr"] .test2 {
    margin-right: 5px;
}

[dir="rtl"] .test1, [dir="rtl"] .test2 {
    margin-left: 5px;
}
```

Expected Output

```
[dir="ltr"] .test1, [dir="ltr"] .test2 {
    margin-right: 5px;
}

[dir="rtl"] .test1, [dir="rtl"] .test2 {
    margin-left: 5px;
}
```

Or perhaps


Expected Output

```
[dir="ltr"] .test1, [dir="ltr"] .test2 {
    margin-right: 4px;
    margin-right: 5px;
}

[dir="rtl"] .test1, [dir="rtl"] .test2 {
    margin-left: 4px;
    margin-left: 5px;
}
```

Thanks for taking the time to understand my use case!

**Comment by elchininet** [comment]
Hi @jimmy-guo,

Yes, I got it before when I noticed that we were talking about declarations that flips their property instead of their value. The bug is already fixed in [version 4.0.2 of the package](https://www.npmjs.com/package/postcss-rtlcss/v/4.0.2). Try it out and let me know your results.

Regards and thanks for helping to improve the plugin 👍
